zboxcore/sdk/downloadworker.go

```diff
@@ -46,338 +45,494 @@ type DownloadRequest struct {
+			req.statusCallback.InProgress(req.allocationID, remotePathCB, OpDownload, downloaded, data)
 		}
-
 		if (startBlock + numBlocks) > endBlock {
 			startBlock += endBlock - startBlock
 		} else {
 			startBlock += numBlocks
 		}
 	}
 
-	// Only check hash when the download request is not by block/partial.
-	if req.endBlock == chunksPerShard && req.startBlock == 0 {
-		//calcHash := fileHasher.GetHash()
-		merkleRoot := fileHasher.GetMerkleRoot()
-
-		expectedHash := fileRef.ActualFileHash
-		if req.contentMode == DOWNLOAD_CONTENT_THUMB {
-			expectedHash = fileRef.ActualThumbnailHash
-		}
-
-		//if calcHash != expectedHash && expectedHash != merkleRoot {
-		if expectedHash != merkleRoot {
-			sys.Files.Remove(req.localpath) //nolint: errcheck
-			if req.statusCallback != nil {
-				req.statusCallback.Error(req.allocationID, remotePathCallback, OpDownload, errors.New("", "File content didn't match with uploaded file"))
-			}
+	if isFullDownload {
+		err := req.checkContentHash(fRef, fileHasher, remotePathCB)
+		if err != nil {
+			logger.Logger.Error(err)
 			return
 		}
 	}
 
-	wrFile.Sync()
-	wrFile.Close()
-	wrFile, _ = sys.Files.Open(req.localpath)
-	defer wrFile.Close()
-	wrFile.Seek(0, 0)
-	mimetype, _ := zboxutil.GetFileContentType(wrFile)
+	f.Sync()
 	if req.statusCallback != nil {
-		req.statusCallback.Completed(req.allocationID, remotePathCallback, fileRef.Name, mimetype, int(fileRef.ActualFileSize), OpDownload)
+		req.statusCallback.Completed(
+			req.allocationID, remotePathCB, fRef.Name, "", int(fRef.ActualFileSize), OpDownload)
+	}
+}
+
+func (req *DownloadRequest) initEC() error {
+	var err error
+	req.ecEncoder, err = reedsolomon.New(
+		req.datashards, req.parityshards,
+		reedsolomon.WithAutoGoroutines(int(req.effectiveChunkSize)))
+
+	if err != nil {
+		return errors.New("init_ec",
+			fmt.Sprintf("Got error %s, while initializing erasure encoder", err.Error()))
+	}
+	return nil
+}
+
+func (req *DownloadRequest) initEncryption() {
+	req.encScheme = encryption.NewEncryptionScheme()
+	req.encScheme.Initialize(client.GetClient().Mnemonic)
+	req.encScheme.InitForDecryption("filetype:audio", req.encryptedKey)
+}
+
+func (req *DownloadRequest) errorCB(err error, remotePathCB string) {
+	sys.Files.Remove(req.localpath) //nolint: errcheck
+	if req.statusCallback != nil {
+		req.statusCallback.Error(
+			req.allocationID, remotePathCB, OpDownload, err)
+	}
+	return
+}
+
+func (req *DownloadRequest) checkContentHash(
+	fRef *fileref.FileRef, fileHasher *downloadHasher, remotepathCB string) (err error) {
+
+	hash := fileHasher.GetHash()
+	expectedHash := fRef.ActualFileHash
+	if req.contentMode == DOWNLOAD_CONTENT_THUMB {
+		expectedHash = fRef.ActualThumbnailHash
 	}
+
+	if expectedHash != hash {
+		err = errors.New("merkle_root_mismatch", "File content didn't match with uploaded file")
+		req.errorCB(err, remotepathCB)
+		return
+	}
+	return nil
+}
+
+func (req *DownloadRequest) getFileHandler(remotePathCB string) (sys.File, error) {
```

Comment:
Rename to openFile, the Handler may confuse users and would be considered as a http handler

Reply:
`getFileHandler` should be clear enough I think.
If you insist, I will change it.

Reply:
Yes, please change it. We don't use `handler` in go often to represent a file struct. It might be common in Java, C/C++, but let's keep it as file as the go standard code does.